<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html><head><title>Python: module papillon</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
</head><body bgcolor="#f0f0f8">

<table width="100%" cellspacing=0 cellpadding=2 border=0 summary="heading">
<tr bgcolor="#7799ee">
<td valign=bottom>&nbsp;<br>
<font color="#ffffff" face="helvetica, arial">&nbsp;<br><big><big><strong>papillon</strong></big></big></font></td
><td align=right valign=bottom
><font color="#ffffff" face="helvetica, arial"><a href=".">index</a><br><a href="file:/home/lab/Documents/GitHub/Papillon/papillon.py">/home/lab/Documents/GitHub/Papillon/papillon.py</a></font></td></tr></table>
    <p><tt>A&nbsp;python&nbsp;version&nbsp;of&nbsp;CummeRbund&nbsp;to&nbsp;read&nbsp;and&nbsp;plot&nbsp;Galaxy/cuffdiff&nbsp;RNA-seq&nbsp;data<br>
&nbsp;<br>
Before&nbsp;To&nbsp;start<br>
========<br>
&nbsp;<br>
After&nbsp;RNA-seq&nbsp;analysis&nbsp;with&nbsp;Galaxy,<br>
download&nbsp;the&nbsp;4&nbsp;files&nbsp;generated&nbsp;by&nbsp;cuffdiff&nbsp;containing&nbsp;respectively:<br>
&nbsp;<br>
...&nbsp;transcript_FPKM_tracking...<br>
&nbsp;<br>
...&nbsp;gene_FPKM_tracking&nbsp;...<br>
&nbsp;<br>
...&nbsp;gene_differential_expression...<br>
&nbsp;<br>
...&nbsp;transcript_differential_expression...<br>
&nbsp;<br>
And&nbsp;put&nbsp;them&nbsp;in&nbsp;the&nbsp;same&nbsp;folder.<br>
&nbsp;<br>
You&nbsp;can&nbsp;either&nbsp;use&nbsp;directly&nbsp;the&nbsp;files&nbsp;or&nbsp;change&nbsp;the&nbsp;names&nbsp;according&nbsp;to<br>
cummeRbund&nbsp;in:<br>
&nbsp;<br>
...&nbsp;transcript\_FPKM\_tracking...&nbsp;=&nbsp;isoforms.fpkm_tracking<br>
&nbsp;<br>
...&nbsp;gene\_FPKM\_tracking&nbsp;...&nbsp;=&nbsp;gene.fpkm_tracking<br>
&nbsp;<br>
...&nbsp;gene\_differential\_expression...&nbsp;=&nbsp;gene_exp.diff<br>
&nbsp;<br>
...&nbsp;transcript\_differential\_expression...&nbsp;=&nbsp;isoform_exp.diff</tt></p>
<p>
<table width="100%" cellspacing=0 cellpadding=2 border=0 summary="section">
<tr bgcolor="#aa55cc">
<td colspan=3 valign=bottom>&nbsp;<br>
<font color="#ffffff" face="helvetica, arial"><big><strong>Modules</strong></big></font></td></tr>
    
<tr><td bgcolor="#aa55cc"><tt>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</tt></td><td>&nbsp;</td>
<td width="100%"><table width="100%" summary="list"><tr><td width="25%" valign=top><a href="os.html">os</a><br>
<a href="pandas.html">pandas</a><br>
</td><td width="25%" valign=top><a href="re.html">re</a><br>
<a href="seaborn.html">seaborn</a><br>
</td><td width="25%" valign=top><a href="warnings.html">warnings</a><br>
</td><td width="25%" valign=top></td></tr></table></td></tr></table><p>
<table width="100%" cellspacing=0 cellpadding=2 border=0 summary="section">
<tr bgcolor="#ee77aa">
<td colspan=3 valign=bottom>&nbsp;<br>
<font color="#ffffff" face="helvetica, arial"><big><strong>Classes</strong></big></font></td></tr>
    
<tr><td bgcolor="#ee77aa"><tt>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</tt></td><td>&nbsp;</td>
<td width="100%"><dl>
<dt><font face="helvetica, arial"><a href="builtins.html#object">builtins.object</a>
</font></dt><dd>
<dl>
<dt><font face="helvetica, arial"><a href="papillon.html#Papillon_db">Papillon_db</a>
</font></dt><dt><font face="helvetica, arial"><a href="papillon.html#Papillon_list">Papillon_list</a>
</font></dt></dl>
</dd>
</dl>
 <p>
<table width="100%" cellspacing=0 cellpadding=2 border=0 summary="section">
<tr bgcolor="#ffc8d8">
<td colspan=3 valign=bottom>&nbsp;<br>
<font color="#000000" face="helvetica, arial"><a name="Papillon_db">class <strong>Papillon_db</strong></a>(<a href="builtins.html#object">builtins.object</a>)</font></td></tr>
    
<tr bgcolor="#ffc8d8"><td rowspan=2><tt>&nbsp;&nbsp;&nbsp;</tt></td>
<td colspan=2><tt>Create&nbsp;a&nbsp;<a href="#Papillon_db">Papillon_db</a>&nbsp;<a href="builtins.html#object">object</a>&nbsp;using&nbsp;<a href="#-read_folder">read_folder</a>()&nbsp;or&nbsp;<a href="#-read_files">read_files</a>()&nbsp;and&nbsp;_papillon_builder<br>&nbsp;</tt></td></tr>
<tr><td>&nbsp;</td>
<td width="100%">Methods defined here:<br>
<dl><dt><a name="Papillon_db-__init__"><strong>__init__</strong></a>(self, path, samples, comparisons, genes_detected, isoforms_detected)</dt><dd><tt>self.<strong>path</strong>&nbsp;-&nbsp;files&nbsp;path<br>
self.<strong>samples</strong>&nbsp;-&nbsp;samples&nbsp;found<br>
self.<strong>comparisons</strong>&nbsp;-&nbsp;comparisons&nbsp;found<br>
self.<strong>genes_detected</strong>&nbsp;-&nbsp;dataframe&nbsp;of&nbsp;genes&nbsp;detected<br>
self.<strong>isoforms_detected</strong>&nbsp;-&nbsp;dataframe&nbsp;of&nbsp;isoforms&nbsp;detected</tt></dd></dl>

<dl><dt><a name="Papillon_db-__str__"><strong>__str__</strong></a>(self)</dt><dd><tt>Return&nbsp;str(self).</tt></dd></dl>

<dl><dt><a name="Papillon_db-change_order"><strong>change_order</strong></a>(self, new_order)</dt><dd><tt>Change&nbsp;the&nbsp;samples&nbsp;order<br>
&nbsp;<br>
Parameters<br>
----------<br>
new_order:&nbsp;list&nbsp;of&nbsp;samples&nbsp;order<br>
&nbsp;<br>
Example<br>
----------<br>
MyProject.<a href="#Papillon_db-change_order">change_order</a>(["Sample&nbsp;4","Sample&nbsp;3","Sample&nbsp;2","Sample&nbsp;1"])</tt></dd></dl>

<dl><dt><a name="Papillon_db-drop_comparison"><strong>drop_comparison</strong></a>(self, comparison)</dt><dd><tt>Drop&nbsp;Comparison&nbsp;(str)&nbsp;or&nbsp;list&nbsp;of&nbsp;comparisons<br>
&nbsp;<br>
Parameters<br>
----------<br>
comparison:&nbsp;comparison&nbsp;(str)&nbsp;or&nbsp;list&nbsp;of&nbsp;comparisons<br>
&nbsp;<br>
Example<br>
----------<br>
MyProject.<a href="#Papillon_db-drop_comparison">drop_comparison</a>(specific_comparison)</tt></dd></dl>

<dl><dt><a name="Papillon_db-get_gene"><strong>get_gene</strong></a>(self, genelist=None, comparison='all', comparison_sign=None, fold_ind=None, fold_sign='&gt;')</dt><dd><tt>This&nbsp;method&nbsp;select&nbsp;genes&nbsp;per&nbsp;name&nbsp;or&nbsp;conditions.&nbsp;It&nbsp;return&nbsp;a&nbsp;<a href="#Papillon_list">Papillon_list</a>&nbsp;<a href="builtins.html#object">object</a><br>
&nbsp;<br>
Parameters<br>
----------<br>
genelist&nbsp;-&nbsp;accept&nbsp;string&nbsp;(1&nbsp;gene&nbsp;name),&nbsp;list&nbsp;of&nbsp;gene&nbsp;names&nbsp;or&nbsp;file&nbsp;with&nbsp;a&nbsp;list&nbsp;of&nbsp;gene&nbsp;names&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<br>
comparison&nbsp;-&nbsp;To&nbsp;select&nbsp;genes&nbsp;higher/lower&nbsp;in&nbsp;one&nbsp;condition&nbsp;compared&nbsp;to&nbsp;another.&nbsp;Accept&nbsp;either&nbsp;"all"&nbsp;to&nbsp;pass&nbsp;all&nbsp;the&nbsp;comparisons,&nbsp;or&nbsp;accept&nbsp;only&nbsp;1&nbsp;comparison&nbsp;as&nbsp;str&nbsp;(already&nbsp;present&nbsp;in&nbsp;the&nbsp;data)<br>
comparison_sign&nbsp;-&nbsp;usable&nbsp;in&nbsp;combination&nbsp;with&nbsp;comparison,&nbsp;accept&nbsp;either&nbsp;"&gt;"&nbsp;or&nbsp;"&lt;"<br>
fold_ind&nbsp;-&nbsp;fold&nbsp;induction&nbsp;(log2)&nbsp;higher/lower&nbsp;then&nbsp;number<br>
fold_sign&nbsp;-&nbsp;usable&nbsp;in&nbsp;combination&nbsp;with&nbsp;fold_ind,&nbsp;accept&nbsp;either&nbsp;"&gt;"&nbsp;or&nbsp;"&lt;"<br>
&nbsp;<br>
Example<br>
----------<br>
Selection=MyProject.<a href="#Papillon_db-get_gene">get_gene</a>(["CD44","CCL15"],&nbsp;comparison="Sample&nbsp;1_vs_Sample&nbsp;2",&nbsp;comparison_sign="&lt;",&nbsp;fold_ind=1,&nbsp;fold_sign="&lt;")</tt></dd></dl>

<dl><dt><a name="Papillon_db-get_isoform"><strong>get_isoform</strong></a>(self, genelist=None, comparison='all', comparison_sign=None, fold_ind=None, fold_sign='&gt;')</dt><dd><tt>This&nbsp;function&nbsp;select&nbsp;isoforms.&nbsp;It&nbsp;creates&nbsp;a&nbsp;Papillon&nbsp;<a href="builtins.html#object">object</a><br>
&nbsp;<br>
Parameters<br>
----------<br>
genelist&nbsp;-&nbsp;accept&nbsp;string&nbsp;(1&nbsp;gene&nbsp;name),&nbsp;list&nbsp;of&nbsp;gene&nbsp;names&nbsp;or&nbsp;file&nbsp;with&nbsp;a&nbsp;list&nbsp;of&nbsp;gene&nbsp;names&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<br>
comparison&nbsp;-&nbsp;To&nbsp;select&nbsp;genes&nbsp;higher/lower&nbsp;in&nbsp;one&nbsp;condition&nbsp;compared&nbsp;to&nbsp;another.&nbsp;Accept&nbsp;either&nbsp;"all"&nbsp;to&nbsp;pass&nbsp;all&nbsp;the&nbsp;comparisons,&nbsp;or&nbsp;accept&nbsp;only&nbsp;1&nbsp;comparison&nbsp;as&nbsp;str&nbsp;(already&nbsp;present&nbsp;in&nbsp;the&nbsp;data)<br>
comparison_sign&nbsp;-&nbsp;usable&nbsp;in&nbsp;combination&nbsp;with&nbsp;comparison,&nbsp;accept&nbsp;either&nbsp;"&gt;"&nbsp;or&nbsp;"&lt;"<br>
fold_ind&nbsp;-&nbsp;fold&nbsp;induction&nbsp;(log2)&nbsp;higher/lower&nbsp;then&nbsp;number<br>
fold_sign&nbsp;-&nbsp;usable&nbsp;in&nbsp;combination&nbsp;with&nbsp;fold_ind,&nbsp;accept&nbsp;either&nbsp;"&gt;"&nbsp;or&nbsp;"&lt;"<br>
&nbsp;<br>
Example<br>
----------&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<br>
Selection=MyProject.<a href="#Papillon_db-get_isoform">get_isoform</a>(["IL6","CCL15","IL17RC"],&nbsp;comparison="Sample&nbsp;2_vs_Sample&nbsp;4",&nbsp;comparison_sign="&lt;")</tt></dd></dl>

<dl><dt><a name="Papillon_db-search"><strong>search</strong></a>(self, word, where, how='table', export=False)</dt><dd><tt>search&nbsp;among&nbsp;genes/isoforms&nbsp;names&nbsp;in&nbsp;detected&nbsp;and&nbsp;significant<br>
&nbsp;<br>
Parameters<br>
----------<br>
word&nbsp;-&nbsp;accept&nbsp;a&nbsp;str&nbsp;to&nbsp;search&nbsp;among&nbsp;the&nbsp;gene&nbsp;names<br>
where&nbsp;-&nbsp;accept:<br>
&nbsp;&nbsp;&nbsp;&nbsp;"genes_detected"<br>
&nbsp;&nbsp;&nbsp;&nbsp;"genes_significant"<br>
&nbsp;&nbsp;&nbsp;&nbsp;"isoforms_detected"<br>
&nbsp;&nbsp;&nbsp;&nbsp;"isoforms_significant"<br>
&nbsp;<br>
how&nbsp;-&nbsp;accept:<br>
&nbsp;&nbsp;&nbsp;&nbsp;"table"&nbsp;return&nbsp;the&nbsp;dataframe&nbsp;with&nbsp;the&nbsp;genes&nbsp;found<br>
&nbsp;&nbsp;&nbsp;&nbsp;"list"&nbsp;return&nbsp;a&nbsp;list&nbsp;of&nbsp;names,&nbsp;no&nbsp;duplicates<br>
&nbsp;&nbsp;&nbsp;&nbsp;<br>
export&nbsp;-&nbsp;True/False<br>
&nbsp;&nbsp;&nbsp;&nbsp;<br>
Example<br>
----------<br>
search_result=MyProject.<a href="#Papillon_db-search">search</a>(word="IL6",where="genes_significant",&nbsp;how="table")</tt></dd></dl>

<hr>
Data descriptors defined here:<br>
<dl><dt><strong>__dict__</strong></dt>
<dd><tt>dictionary&nbsp;for&nbsp;instance&nbsp;variables&nbsp;(if&nbsp;defined)</tt></dd>
</dl>
<dl><dt><strong>__weakref__</strong></dt>
<dd><tt>list&nbsp;of&nbsp;weak&nbsp;references&nbsp;to&nbsp;the&nbsp;object&nbsp;(if&nbsp;defined)</tt></dd>
</dl>
</td></tr></table> <p>
<table width="100%" cellspacing=0 cellpadding=2 border=0 summary="section">
<tr bgcolor="#ffc8d8">
<td colspan=3 valign=bottom>&nbsp;<br>
<font color="#000000" face="helvetica, arial"><a name="Papillon_list">class <strong>Papillon_list</strong></a>(<a href="builtins.html#object">builtins.object</a>)</font></td></tr>
    
<tr bgcolor="#ffc8d8"><td rowspan=2><tt>&nbsp;&nbsp;&nbsp;</tt></td>
<td colspan=2><tt>Class&nbsp;containing&nbsp;a&nbsp;selected&nbsp;list&nbsp;of&nbsp;genes,&nbsp;with&nbsp;data&nbsp;associated,&nbsp;generated&nbsp;by&nbsp;<a href="#Papillon_db">Papillon_db</a>.get_gene()&nbsp;or&nbsp;<a href="#Papillon_db">Papillon_db</a>.get_isoform()&nbsp;methods<br>
&nbsp;<br>
Is&nbsp;possible&nbsp;add&nbsp;the&nbsp;content&nbsp;of&nbsp;two&nbsp;or&nbsp;more&nbsp;<a href="#Papillon_list">Papillon_list</a>&nbsp;objects.<br>
Example:<br>
P_list3&nbsp;=&nbsp;P_list1+P_list2<br>
P_list4=sum([PL1,&nbsp;PL2,&nbsp;PL3])<br>&nbsp;</tt></td></tr>
<tr><td>&nbsp;</td>
<td width="100%">Methods defined here:<br>
<dl><dt><a name="Papillon_list-__add__"><strong>__add__</strong></a>(self, other)</dt></dl>

<dl><dt><a name="Papillon_list-__init__"><strong>__init__</strong></a>(self, df, what, comparisons, path, samples, comparison='all', comparison_sign=None, fold_ind=None, fold_sign='&gt;', p=0.05)</dt><dd><tt>Initialize&nbsp;self.&nbsp;&nbsp;See&nbsp;help(type(self))&nbsp;for&nbsp;accurate&nbsp;signature.</tt></dd></dl>

<dl><dt><a name="Papillon_list-__radd__"><strong>__radd__</strong></a>(self, other)</dt></dl>

<dl><dt><a name="Papillon_list-__str__"><strong>__str__</strong></a>(self)</dt><dd><tt>Return&nbsp;str(self).</tt></dd></dl>

<dl><dt><a name="Papillon_list-compare"><strong>compare</strong></a>(self, other)</dt><dd><tt>Compare&nbsp;two&nbsp;<a href="#Papillon_list">Papillon_list</a>&nbsp;objects<br>
&nbsp;<br>
Parameters<br>
----------<br>
other:&nbsp;another&nbsp;<a href="#Papillon_list">Papillon_list</a>&nbsp;<a href="builtins.html#object">object</a><br>
&nbsp;<br>
Example<br>
----------<br>
Selection1.<a href="#Papillon_list-compare">compare</a>(Selection2)</tt></dd></dl>

<dl><dt><a name="Papillon_list-export"><strong>export</strong></a>(self)</dt><dd><tt>Export&nbsp;the&nbsp;selected&nbsp;genes/isoforms&nbsp;as&nbsp;excel&nbsp;file.<br>
&nbsp;<br>
Example<br>
----------<br>
Selection.<a href="#Papillon_list-export">export</a>()</tt></dd></dl>

<dl><dt><a name="Papillon_list-heatmap"><strong>heatmap</strong></a>(self, z_score=False, col_cluster=False, method='complete', cmap='seismic', export=False, **options)</dt><dd><tt>Generate&nbsp;heatmap&nbsp;using&nbsp;selected&nbsp;genes/isoforms<br>
&nbsp;<br>
Parameters<br>
----------<br>
z_score&nbsp;-&nbsp;True/False&nbsp;whether&nbsp;want&nbsp;or&nbsp;not&nbsp;apply&nbsp;z-score&nbsp;normalization<br>
col_cluster&nbsp;-&nbsp;True/False&nbsp;whether&nbsp;want&nbsp;or&nbsp;not&nbsp;cluster&nbsp;the&nbsp;samples<br>
method&nbsp;-&nbsp;clustering&nbsp;algorithm&nbsp;-&nbsp;default&nbsp;is&nbsp;complete-linkage<br>
cmap&nbsp;-&nbsp;map&nbsp;color<br>
export&nbsp;-&nbsp;True/False&nbsp;whether&nbsp;want&nbsp;or&nbsp;not&nbsp;export&nbsp;the&nbsp;dataframe&nbsp;of<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;selected&nbsp;genes<br>
**options&nbsp;-&nbsp;all&nbsp;the&nbsp;options&nbsp;accepted&nbsp;by&nbsp;seaborn.clustermap<br>
default&nbsp;metric&nbsp;is&nbsp;euclidean.<br>
&nbsp;<br>
Example<br>
----------<br>
Selection.<a href="#Papillon_list-heatmap">heatmap</a>(z_score=True,export=True)</tt></dd></dl>

<dl><dt><a name="Papillon_list-lineplot"><strong>lineplot</strong></a>(self, title='', legend=True, z_score=False, export=False, size=10, ci=None, **option)</dt><dd><tt>LinePlot&nbsp;selected&nbsp;genes&nbsp;expression&nbsp;levels.&nbsp;Max&nbsp;number&nbsp;of&nbsp;genes&nbsp;200<br>
&nbsp;<br>
Parameters<br>
----------<br>
title&nbsp;-&nbsp;accept&nbsp;a&nbsp;str&nbsp;as&nbsp;title&nbsp;of&nbsp;the&nbsp;plot<br>
legend&nbsp;-&nbsp;True/False&nbsp;show&nbsp;the&nbsp;legend<br>
z_score&nbsp;-&nbsp;True/False&nbsp;calculate&nbsp;the&nbsp;z-score&nbsp;normalization<br>
export&nbsp;-&nbsp;True/False&nbsp;whether&nbsp;or&nbsp;not&nbsp;export&nbsp;the&nbsp;image<br>
size&nbsp;-&nbsp;change&nbsp;the&nbsp;size&nbsp;of&nbsp;the&nbsp;plot<br>
**options&nbsp;-&nbsp;all&nbsp;the&nbsp;options&nbsp;accepted&nbsp;by&nbsp;seaborn.factorplot<br>
&nbsp;<br>
Example<br>
----------<br>
Selection.<a href="#Papillon_list-lineplot">lineplot</a>(export=True,z_score=True)</tt></dd></dl>

<dl><dt><a name="Papillon_list-onlyFPKM"><strong>onlyFPKM</strong></a>(self, return_as, remove_FPKM_name=False)</dt><dd><tt>Take&nbsp;a&nbsp;<a href="#Papillon_list">Papillon_list</a>&nbsp;<a href="builtins.html#object">object</a>&nbsp;and&nbsp;return&nbsp;only&nbsp;FPKM&nbsp;columns.<br>
&nbsp;<br>
Parameters<br>
----------<br>
return&nbsp;as:<br>
&nbsp;&nbsp;&nbsp;&nbsp;"df"&nbsp;-&nbsp;pandas&nbsp;DataFrame<br>
&nbsp;&nbsp;&nbsp;&nbsp;"array"&nbsp;-&nbsp;numpy&nbsp;array<br>
&nbsp;&nbsp;&nbsp;&nbsp;"gene&nbsp;name"&nbsp;-&nbsp;pandas&nbsp;DataFrame&nbsp;containing&nbsp;gene&nbsp;names<br>
&nbsp;<br>
&nbsp;<br>
remove_FPKM_name:&nbsp;True/False<br>
&nbsp;<br>
Example<br>
----------<br>
df=Selection.<a href="#Papillon_list-onlyFPKM">onlyFPKM</a>("df")</tt></dd></dl>

<dl><dt><a name="Papillon_list-plot"><strong>plot</strong></a>(**parameter)</dt><dd><tt>Deprecated.&nbsp;Use&nbsp;self.<a href="#Papillon_list-lineplot">lineplot</a>()&nbsp;instead</tt></dd></dl>

<dl><dt><a name="Papillon_list-search"><strong>search</strong></a>(self, string)</dt><dd><tt>Search&nbsp;a&nbsp;string&nbsp;in&nbsp;the&nbsp;<a href="#Papillon_list">Papillon_list</a><br>
&nbsp;<br>
Parameters<br>
----------<br>
string&nbsp;-&nbsp;accept&nbsp;a&nbsp;str&nbsp;to&nbsp;search&nbsp;among&nbsp;the&nbsp;gene&nbsp;names<br>
&nbsp;<br>
Example<br>
----------<br>
search_results=Selection.<a href="#Papillon_list-search">search</a>("IL")</tt></dd></dl>

<dl><dt><a name="Papillon_list-select"><strong>select</strong></a>(self, genelist)</dt><dd><tt>Create&nbsp;another&nbsp;<a href="#Papillon_list">Papillon_list</a>&nbsp;<a href="builtins.html#object">object</a><br>
&nbsp;<br>
Parameters<br>
----------<br>
genelist:&nbsp;accept&nbsp;string&nbsp;(1&nbsp;gene&nbsp;name),&nbsp;list&nbsp;of&nbsp;gene&nbsp;names&nbsp;or&nbsp;file<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;with&nbsp;a&nbsp;list&nbsp;of&nbsp;gene&nbsp;names<br>
&nbsp;<br>
Example<br>
----------<br>
Selection2=Selection1.<a href="#Papillon_list-select">select</a>(["IL6","CCL15"])</tt></dd></dl>

<dl><dt><a name="Papillon_list-show"><strong>show</strong></a>(self)</dt><dd><tt>Show&nbsp;genes/isoforms&nbsp;as&nbsp;Dataframe<br>
&nbsp;<br>
Example<br>
----------<br>
Selection.<a href="#Papillon_list-show">show</a>()</tt></dd></dl>

<hr>
Data descriptors defined here:<br>
<dl><dt><strong>__dict__</strong></dt>
<dd><tt>dictionary&nbsp;for&nbsp;instance&nbsp;variables&nbsp;(if&nbsp;defined)</tt></dd>
</dl>
<dl><dt><strong>__weakref__</strong></dt>
<dd><tt>list&nbsp;of&nbsp;weak&nbsp;references&nbsp;to&nbsp;the&nbsp;object&nbsp;(if&nbsp;defined)</tt></dd>
</dl>
</td></tr></table></td></tr></table><p>
<table width="100%" cellspacing=0 cellpadding=2 border=0 summary="section">
<tr bgcolor="#eeaa77">
<td colspan=3 valign=bottom>&nbsp;<br>
<font color="#ffffff" face="helvetica, arial"><big><strong>Functions</strong></big></font></td></tr>
    
<tr><td bgcolor="#eeaa77"><tt>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</tt></td><td>&nbsp;</td>
<td width="100%"><dl><dt><a name="-read_db"><strong>read_db</strong></a>(path, drop_comparison=None)</dt><dd><tt>Deprecated.&nbsp;Use&nbsp;<a href="#-read_folder">read_folder</a>()&nbsp;instead.</tt></dd></dl>
 <dl><dt><a name="-read_files"><strong>read_files</strong></a>(files, path=None, drop_comparison=None)</dt><dd><tt>Accept&nbsp;cuffdiff/cummeRbund&nbsp;files&nbsp;as&nbsp;iterable<br>
&nbsp;&nbsp;&nbsp;&nbsp;("transcript_FPKM_tracking",<br>
&nbsp;&nbsp;&nbsp;&nbsp;"gene_FPKM_tracking",<br>
&nbsp;&nbsp;&nbsp;&nbsp;"gene_differential_expression",<br>
&nbsp;&nbsp;&nbsp;&nbsp;"transcript_differential_expression")<br>
and&nbsp;return&nbsp;them&nbsp;to&nbsp;_papillon_builder()&nbsp;to&nbsp;create&nbsp;a&nbsp;<a href="#Papillon_db">Papillon_db</a>&nbsp;<a href="builtins.html#object">object</a>.<br>
&nbsp;<br>
Parameters<br>
----------<br>
&nbsp;<br>
files&nbsp;-&nbsp;accept&nbsp;an&nbsp;iterable&nbsp;with&nbsp;the&nbsp;cuffdiff&nbsp;files&nbsp;names<br>
path&nbsp;-&nbsp;where&nbsp;to&nbsp;export&nbsp;Papillon&nbsp;generated&nbsp;files<br>
drop_comparison&nbsp;-&nbsp;drop&nbsp;comparison&nbsp;(str)&nbsp;or&nbsp;list&nbsp;of&nbsp;comparisons&nbsp;to&nbsp;drop&nbsp;from&nbsp;the&nbsp;cuffdiff&nbsp;table<br>
&nbsp;<br>
Example<br>
----------<br>
pp.<a href="#-read_files">read_files</a>(["Files/gene_exp.diff","Files/genes.fpkm_tracking","Files/isoform_exp.diff","Files/isoforms.fpkm_tracking"])</tt></dd></dl>
 <dl><dt><a name="-read_folder"><strong>read_folder</strong></a>(path, drop_comparison=None)</dt><dd><tt>Read&nbsp;the&nbsp;folder&nbsp;containing&nbsp;the&nbsp;cuffdiff/cummeRbund&nbsp;files,&nbsp;and&nbsp;return<br>
them&nbsp;to&nbsp;_papillon_builder()&nbsp;to&nbsp;create&nbsp;a&nbsp;<a href="#Papillon_db">Papillon_db</a>&nbsp;<a href="builtins.html#object">object</a>.<br>
&nbsp;<br>
&nbsp;<br>
Parameters<br>
----------<br>
path&nbsp;-&nbsp;accept&nbsp;a&nbsp;str&nbsp;with&nbsp;the&nbsp;folder&nbsp;path,&nbsp;containing&nbsp;the&nbsp;cuffdiff&nbsp;files<br>
drop_comparison&nbsp;-&nbsp;drop&nbsp;comparison&nbsp;(str)&nbsp;or&nbsp;list&nbsp;of&nbsp;comparisons&nbsp;to&nbsp;drop&nbsp;from&nbsp;the&nbsp;cuffdiff&nbsp;table<br>
&nbsp;<br>
Example<br>
----------<br>
MyProject=pp.<a href="#-read_folder">read_folder</a>("MyFolder/Test_files")</tt></dd></dl>
</td></tr></table>
</body></html>